### PR TITLE
feat: add C implementation for `stats/base/dists/uniform/median`

### LINKED ISSUE #3831: [RFC]: Add C implementation for `@stdlib/stats/base/dists/uniform/median`
### Description

This RFC proposes adding a C implementation, including a Node.js native addon, C benchmarks, and C examples, for the [`@stdlib/stats/base/dists/uniform/median`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/@stdlib/stats/base/dists/uniform/median) package.

When adding support, the following tasks should be completed:

1. A C source implementation should be added to a `src` folder within the package's directory.
2. A native add-on interface should be added in order to call the C implementation from JavaScript.
3. A JavaScript file should be added to the `lib` folder to allow benchmarking and unit testing the native implementation from JavaScript.
4. A new test file should be created which includes unit tests testing against expected behavior.
5. C benchmarks should be added which measure performance for the C source implementation.
6. A file containing C examples should be added to demonstrate example usage.
7. JavaScript benchmarks should be added which measure performance for the native add-on.
8. Build and configuration files should be added to allow compiling and running the C source implementation using project `make` commands (as documented below).
9. The README should be updated to include documentation for the added C API.

To provide a concrete example of what a PR adding the desired method should contain, see https://github.com/stdlib-js/stdlib/pull/3354, which is a PR adding a C implementation for the CDF of an arcsine distribution. This should provide an idea of what is expected.

#### Prerequisites

1. Review JavaScript implementation in `lib/main.js` and its required modules to identify all `stdlib` functions used.
2. Verify C implementations exist for each required function.

### Related Issues

#3355

### Questions

No.

### Other

Once the implementation is ready including C examples and benchmarks, we should be able to run the following `make` commands.

#### Build native add-on

```bash
NODE_ADDONS_PATTERN="@stdlib/stats/base/dists/uniform/median" make install-node-addons
```

#### Run C examples

```bash
make examples-c-files FILES="$(pwd)/lib/node_modules/@stdlib/stats/base/dists/uniform/median/examples/c/example.c"
```

#### Run C benchmarks

```bash
make benchmark-c-files FILES="$(pwd)/lib/node_modules/@stdlib/stats/base/dists/uniform/median/benchmark/c/benchmark.c"
```

#### Run JavaScript benchmarks

```bash
make benchmark-javascript-files FILES="$(pwd)/lib/node_modules/@stdlib/stats/base/dists/uniform/median/benchmark/benchmark.native.js"
```

#### Run all tests

```bash
make TESTS_FILTER=".*/@stdlib/stats/base/dists/uniform/median/.*" test
```

### Notes

- If you are interested in contributing a PR which addresses this RFC and are a **first-time contributor** or not yet familiar with our project conventions, please minimize the submission of LLM-generated code. Please consult our [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md) and the associated [development guide](https://github.com/stdlib-js/stdlib/blob/develop/docs/development.md). Failure to respect project conventions will result in your PR being rejected without review. Thank you for understanding!

### Checklist

- [x] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [x] Searched for existing issues and pull requests.
- [x] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by 0PrashantYadav0**
Resolves #3831

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for stats/base/dists/uniform/median along with relevant tests, docs, examples and benchmarks.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3831

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/README.md [hunk 1]
```diff
@@ -28,14 +28,14 @@ limitations under the License.
...
-\mathop{\mathrm{Median}}\left[ X \right] = \frac{1}{2} \left( a + b \right)
+\operatorname{median}\left[ X \right] = 0.5 \cdot ( a + b )
 ```
 
-<!-- <div class="equation" align="center" data-raw-text="\operatorname{Median}\left[ X \right] = \frac{1}{2} \left( a + b \right)" data-equation="eq:uniform_median">
-    <img src="https://cdn.jsdelivr.net/gh/stdlib-js/stdlib@591cf9d5c3a0cd3c1ceec961e5c49d73a68374cb/lib/node_modules/@stdlib/stats/base/dists/uniform/median/docs/img/equation_uniform_median.svg" alt="Median for a uniform distribution.">
+<!-- <div class="equation" align="center" data-raw-text="\operatorname{median}\left[ X \right] = 0.5 \cdot ( a + b )" data-equation="eq:uniform_median">
+    <img src="..." alt="Median for a uniform distribution.">
     <br>
 </div> -->
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/README.md]: These changes should be reverted.

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/include/stdlib/stats/base/dists/uniform/median.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/*
+* If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+*/
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* Returns the median of a Uniform distribution with lower bound `a` and upper bound `b`.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/include/stdlib/stats/base/dists/uniform/median.h]:
```suggestion
* Returns the median of a uniform distribution with lower bound `a` and upper bound `b`.
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/lib/native.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+// MODULES //
+
+var addon = require( './../src/addon.node' );
+
+
+// MAIN //
+
+/**
+* Returns the median of a Uniform distribution.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/lib/native.js]:
```suggestion
* Returns the median of a uniform distribution.
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/examples/c/Makefile [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+SOURCE_FILES ?=
+
+# List of libraries (e.g., `-lopenblas -lpthread`):
+LIBRARIES ?=
+
+# List of library paths (e.g., `-L /foo/bar -L /beep/boop`):
+LIBPATH ?=
+
+# List of C targets:
+c_targets := benchmark.out
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/examples/c/Makefile]: The Makefiles in `benchmark/c`, `exampkes/c`, and `src` all have different targets. Please check the reference PR.

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/manifest.json [hunk 5]
```diff
@@ -0,0 +1,76 @@
+{
+    "options": {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/manifest.json]: Should have two-space indentation.

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/src/main.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "stdlib/stats/base/dists/uniform/median.h"
+#include "stdlib/math/base/assert/is_nan.h"
+
+/**
+* Returns the median of a Uniform distribution.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/src/main.c]:
```suggestion
* Returns the median of a uniform distribution.
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/src/main.c [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+* double y = stdlib_base_dists_uniform_median( 2.0, 8.0 );
+* // returns 5.0
+*/
+double stdlib_base_dists_uniform_median( const double a, const double b ) {
+	if (
+		stdlib_base_is_nan( a ) ||
+		stdlib_base_is_nan( b ) ||
+		a >= b
+	) {
+		return 0.0 / 0.0; // Return NaN for invalid inputs
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/src/main.c]:
```suggestion
		return 0.0 / 0.0; // NaN
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/test/test.native.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	y = median( NINF, NINF );
+	t.equal( isnan( y ), true, 'returns NaN' );
+
+	y = median( PINF, NINF );
+	t.equal( isnan( y ), true, 'returns NaN' );
+
+	t.end();
+});
+
+tape( 'the function returns the median of a Uniform distribution', opts, function test( t ) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/test/test.native.js]:
```suggestion
tape( 'the function returns the median of a uniform distribution', opts, function test( t ) {
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/benchmark/benchmark.js [hunk 9]
```diff
@@ -1,7 +1,7 @@
 /**
 * @license Apache-2.0
 *
-* Copyright (c) 2018 The Stdlib Authors.
+* Copyright (c) 2024 The Stdlib Authors.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/benchmark/benchmark.js]: We don't update copyright years when changing existing files.

```suggestion
* Copyright (c) 2018 The Stdlib Authors.
```

**Review by Planeshifter — CHANGES_REQUESTED** [review]
Thanks for your efforts!

The PR currently has a bunch of errors and does not compile. Please test the commands in the linked issue to confirm things work as expected. You also want to make sure you have EditorConfig setup.

**Comment by 0PrashantYadav0** [comment]
@Planeshifter Make all the changes. Please review all the changes.

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/README.md [hunk 10]
```diff
@@ -133,6 +133,96 @@ for ( i = 0; i < 10; i++ ) {
...
+
+### Usage
+
+```c
+#include "stdlib/stats/base/dists/uniform/median.h"
+```
+
+#### stdlib_base_dists_uniform_median( a, b )
+
+Returns the [median][median] of a [uniform][uniform-distribution] distribution with minimum value \( a \) and maximum value \( b \).
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/README.md]:
```suggestion
Returns the [median][median] of a [uniform][uniform-distribution] distribution with parameters `a` (minimum support) and `b` (maximum support).
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/README.md [hunk 11]
```diff
@@ -133,6 +133,96 @@ for ( i = 0; i < 10; i++ ) {
...
+
+```c
+double out = stdlib_base_dists_uniform_median( 0.0, 1.0 );
+// returns 0.5
+```
+
+The function accepts the following arguments:
+
+-   **a**: `[in] double` minimum value.
+-   **b**: `[in] double` maximum value.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/README.md]:
```suggestion
-   **a**: `[in] double` minimum support.
-   **b**: `[in] double` maximum support.
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/src/main.c [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+*/
+
+#include "stdlib/stats/base/dists/uniform/median.h"
+#include "stdlib/math/base/assert/is_nan.h"
+
+/**
+* Returns the median of a uniform distribution.
+*
+* @param a    lower bound
+* @param b    upper bound
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/src/main.c]:
```suggestion
* @param a    minimum support
* @param b    maximum support
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/lib/native.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+
+// MAIN //
+
+/**
+* Returns the median of a uniform distribution.
+*
+* @private
+* @param {number} a - lower bound
+* @param {number} b - upper bound
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/lib/native.js]:
```suggestion
* @param {number} a - minimum support
* @param {number} b - maximum support
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/include/stdlib/stats/base/dists/uniform/median.h [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+/*
+* If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+*/
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* Returns the median of a uniform distribution with lower bound `a` and upper bound `b`.
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/include/stdlib/stats/base/dists/uniform/median.h]:
```suggestion
* Returns the median of a uniform distribution.
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/examples/c/example.c [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+	double a;
+	double b;
+	double y;
+	int i;
+
+	for ( i = 0; i < 25; i++ ) {
+		a = (double)rand() / ( (double)RAND_MAX + 1.0 ) * 100.0;
+		b = a + ( (double)rand() / ( (double)RAND_MAX + 1.0 ) * 50.0 );
+		y = stdlib_base_dists_uniform_median( a, b );
+		printf( "a: %lf , b: %lf , Median: %lf\n", a, b, y );
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/examples/c/example.c]:
```suggestion
		printf( "a: %lf , b: %lf , Median(X;a,b): %lf\n", a, b, y );
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/README.md [hunk 16]
```diff
@@ -133,6 +133,96 @@ for ( i = 0; i < 10; i++ ) {
...
+    double y;
+    int i;
+
+    for ( i = 0; i < 25; i++ ) {
+        a = (double)( rand() % 10 );
+        b = a + (double)( rand() % 10 + 1 ); // ensure b > a
+        y = stdlib_base_dists_uniform_median( a, b );
+        printf( "a: %lf, b: %lf, Median(X;a,b): %lf\n", a, b, y );
+    }
+}
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/README.md]:
```suggestion
static double random_uniform( const double min, const double max ) {
    double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
    return min + ( v*(max-min) );
}

int main( void ) {
    double a;
    double b;
    double y;
    int i;

    for ( i = 0; i < 25; i++ ) {
        a = random_uniform( 0.0, 10.0 );
        b = random_uniform( a, 20.0 ); // ensure b > a
        y = stdlib_base_dists_uniform_median( a, b );
        printf( "a: %lf, b: %lf, Median(X;a,b): %lf\n", a, b, y );
    }
}
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/benchmark/c/benchmark.c [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+			printf( "should not return NaN\n" );
+			break;
+		}
+	}
+	elapsed = tic() - t;
+	if ( y != y ) {
+		printf( "should not return NaN\n" );
+	}
+	return elapsed;
+}
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/benchmark/c/benchmark.c]:
```suggestion

/**
* Generates a random number on the interval [min,max).
*
* @param min    minimum value (inclusive)
* @param max    maximum value (exclusive)
* @return       random number
*/
static double random_uniform( const double min, const double max ) {
	double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
	return min + ( v*(max-min) );
}

/**
* Runs a benchmark.
*
* @return elapsed time in seconds
*/
static double benchmark( void ) {
	double elapsed;
	double a, b;
	double y;
	double t;
	int i;

	t = tic();
	for ( i = 0; i < ITERATIONS; i++ ) {
		a = random_uniform( 0.0, 100.0 );
		b = random_uniform( a, 200.0 );
		y = stdlib_base_dists_uniform_median( a, b );
		if ( y != y ) {
			printf( "should not return NaN\n" );
			break;
		}
	}
	elapsed = tic() - t;
	if ( y != y ) {
		printf( "should not return NaN\n" );
	}
	return elapsed;
}
```

File: lib/node_modules/@stdlib/stats/base/dists/uniform/median/examples/c/example.c [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+int main( void ) {
+	double a;
+	double b;
+	double y;
+	int i;
+
+	for ( i = 0; i < 25; i++ ) {
+		a = (double)rand() / ( (double)RAND_MAX + 1.0 ) * 100.0;
+		b = a + ( (double)rand() / ( (double)RAND_MAX + 1.0 ) * 50.0 );
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/stats/base/dists/uniform/median/examples/c/example.c]:
```suggestion
static double random_uniform( const double min, const double max ) {
	double v = (double)rand() / ( (double)RAND_MAX + 1.0 );
	return min + ( v*(max-min) );
}

int main( void ) {
	double a;
	double b;
	double y;
	int i;

	for ( i = 0; i < 25; i++ ) {
		a = random_uniform( 0.0, 10.0 );
		b = random_uniform( a, 20.0 );
```

**Review by Planeshifter — COMMENTED** [review]
Thanks! 

Made a few edits and fixes. Specifically, we want to be consistent with existing terminology (e.g., use "minimum support" and "maximum support" for parameters of continuous uniform distribution) and use the custom `random_uniform` to draw randon numbers in the example and benchmark code.

**Review by Planeshifter — APPROVED** [review]

**Pull request merged**